### ISSUE TITLE
Parameters wrong when calling C from Zig via a function pointer.

### ISSUE THREAD

**Issue by gruebite** [ziglang/zig#4344]
This gist contains a set of files to reproduce the error. I am creating a struct with 3 pointers and setting them all to null. In the C .so example, the struct and nulls are correctly passed, in Zig they're not. The gist: https://gist.github.com/gruebite/2c2bb1bfb438473d108cc18f717d87f9

I've included the assembly from the gist for the relevant call. My gut says calling conventions are wrong.

Note: I am running on Linux x86_64

Assembly for C
```
=> 0x7ffff7fca138 <init+40>:	mov    -0x28(%rbp),%rax
   0x7ffff7fca13c <init+44>:	mov    (%rax),%rax
   0x7ffff7fca13f <init+47>:	mov    -0x10(%rbp),%rcx
   0x7ffff7fca143 <init+51>:	mov    %rsp,%rdx
   0x7ffff7fca146 <init+54>:	mov    %rcx,0x10(%rdx)
   0x7ffff7fca14a <init+58>:	movaps -0x20(%rbp),%xmm0
   0x7ffff7fca14e <init+62>:	movups %xmm0,(%rdx)
   0x7ffff7fca151 <init+65>:	callq  *%rax
   0x7ffff7fca153 <init+67>:	mov    %fs:0x28,%rax
   0x7ffff7fca15c <init+76>:	mov    -0x8(%rbp),%rcx
   0x7ffff7fca160 <init+80>:	cmp    %rcx,%rax
   0x7ffff7fca163 <init+83>:	jne    0x7ffff7fca16f <init+95>
   0x7ffff7fca169 <init+89>:	add    $0x40,%rsp
   0x7ffff7fca16d <init+93>:	pop    %rbp
   0x7ffff7fca16e <init+94>:	retq   
   0x7ffff7fca16f <init+95>:	callq  0x7ffff7fca030 <__stack_chk_fail@plt>
   0x7ffff7fca174 <_fini>:	endbr64 
   0x7ffff7fca178 <_fini+4>:	sub    $0x8,%rsp
   0x7ffff7fca17c <_fini+8>:	add    $0x8,%rsp
   0x7ffff7fca180 <_fini+12>:	retq   
```

Assembly for Zig

```
=> 0x7ffff7d8f7dd <init+45>:	cmpq   $0x0,-0x20(%rbp)
   0x7ffff7d8f7e2 <init+50>:	jne    0x7ffff7d8f7f4 <init+68>
   0x7ffff7d8f7e4 <init+52>:	xor    %eax,%eax
   0x7ffff7d8f7e6 <init+54>:	mov    %eax,%esi
   0x7ffff7d8f7e8 <init+56>:	lea    0x34bf9(%rip),%rdi        # 0x7ffff7dc43e8 <__unnamed_9>
   0x7ffff7d8f7ef <init+63>:	callq  0x7ffff7d8f000 <std.builtin.default_panic>
   0x7ffff7d8f7f4 <init+68>:	mov    -0x20(%rbp),%rax
   0x7ffff7d8f7f8 <init+72>:	cmpq   $0x0,(%rax)
   0x7ffff7d8f7fc <init+76>:	mov    %rax,-0x28(%rbp)
   0x7ffff7d8f800 <init+80>:	jne    0x7ffff7d8f812 <init+98>
   0x7ffff7d8f802 <init+82>:	xor    %eax,%eax
   0x7ffff7d8f804 <init+84>:	mov    %eax,%esi
   0x7ffff7d8f806 <init+86>:	lea    0x34bdb(%rip),%rdi        # 0x7ffff7dc43e8 <__unnamed_9>
   0x7ffff7d8f80d <init+93>:	callq  0x7ffff7d8f000 <std.builtin.default_panic>
   0x7ffff7d8f812 <init+98>:	lea    -0x18(%rbp),%rdi
   0x7ffff7d8f816 <init+102>:	mov    -0x28(%rbp),%rax
   0x7ffff7d8f81a <init+106>:	callq  *(%rax)
   0x7ffff7d8f81c <init+108>:	add    $0x30,%rsp
   0x7ffff7d8f820 <init+112>:	pop    %rbp
   0x7ffff7d8f821 <init+113>:	retq
```

**Comment by MasterQ32** [comment]
Supplemental code example:
https://gist.github.com/MasterQ32/913cbcfc25e968ad71e52120838d4d17

**Comment by Vexu** [comment]
Fixed by d968d9d10329234afc1d7743bb9981245695ca2b
